File: src/main/scala/kim/dencka/easy/collection/PlusOne.scala
package kim.dencka.easy.collection

object PlusOne extends App {

  def plusOne(digits: Array[Int]): Array[Int] = {
    @annotation.tailrec
    def run(arr: Array[Int], acc: Array[Int], cnt: Int): Array[Int] =
      arr match {
        case Array(x) if x == 9 =>
          println(s"0: acc: ${acc.toList}, arr: ${arr.toList}, cnt: $cnt")
          Array(1, 0).concat(acc)
        case Array(x) =>
          println(s"0: acc: ${acc.toList}, arr: ${arr.toList}, cnt: $cnt")
          Array(x + 1).concat(acc)
        case ar if ar(cnt) == 9 =>
          println(s"1: acc: ${acc.toList}, ar: ${ar.toList}, cnt: $cnt")
          run(ar.init, 0 +: acc, cnt - 1)
        case ar if ar(cnt) != 9 =>
          println(s"2: acc: ${acc.toList}, ar: ${ar.toList}, cnt: $cnt")
          (ar.init :+ ar(cnt) + 1).concat(acc)
      }
    run(digits, Array(), digits.length - 1)
  }

  println(plusOne(Array(9)).toList)
  println(plusOne(Array(1, 2, 3)).toList)
  println(plusOne(Array(9, 9, 9, 9, 9)).toList)
  println(plusOne(Array(8, 9, 9, 9, 9)).toList)
  println(plusOne(Array(5)).toList)

}

File: src/main/scala/kim/dencka/java/easy/collection/PlusOne.java
package kim.dencka.java.easy.collection;

import java.util.Arrays;

class PlusOne {

    // KIM_INFO SHOULD RESOLVE (JAVA)

    public static void main(String[] args) {
        System.out.println(Arrays.toString(plusOne(new int[]{1, 2, 3})));
        System.out.println(Arrays.toString(plusOne(new int[]{4, 3, 2, 1})));
        System.out.println(Arrays.toString(plusOne(new int[]{9})));
    }

    public static int[] plusOne(int[] digits) {
        int n = digits.length;
        int mem;
        if (digits[n - 1] != 9) {
            digits[n - 1] += 1;
            return digits;
        } else {
            mem = 1;
            for (int i = n - 1; i >= 0; i--) {
                if (digits[i] != 9) {
                    digits[i] += mem;
                    mem = 0;
                } else {
                    digits[i] = mem != 0 ? 0 : digits[i];
                }
            }
            if (mem == 0) {
                return digits;
            } else {
                int[] res = new int[n + 1];
                System.arraycopy(digits, 0, res, 1, n);
                res[0] = mem;
                return res;
            }
        }
    }

//    public static int[] plusOne(int[] digits) {
//        String str = "";
//        for (int n : digits) str += n;
//        char[] incrd = (new BigInteger(str).add(new BigInteger("1"))).toString().toCharArray();
//        int[] res = new int[incrd.length];
//        for (int i = 0; i < incrd.length; i++) {
//            res[i] = incrd[i] - '0';
//        }
//        return res;
//    }
}
